<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2024-02-08">

<title>Read processing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./img/UNTRM_favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./img/CABANA_small.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">/STC-CGIAR workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> <i class="bi bi-house" role="img">
</i> 
<span class="menu-text"> </span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-wed-feb-7" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Wed Feb 7</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-wed-feb-7">    
        <li>
    <a class="dropdown-item" href="./01_osc.html">
 <span class="dropdown-text">Supercomputing &amp; OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02_vscode.html">
 <span class="dropdown-text">VS Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03_shell.html">
 <span class="dropdown-text">Unix shell refresher</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./04_fastq.html">
 <span class="dropdown-text">FASTQ files and quality control</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="./05_amplicon.html">
 <span class="dropdown-text"><em>Phytophthora</em> amplicon sequencing</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-thu-feb-8" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Thu Feb 8</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-thu-feb-8">    
        <li class="dropdown-header">Bacterial whole-genome sequencing (slides)</li>
        <li class="dropdown-header">Read preprocessing</li>
        <li class="dropdown-header">Genome assembly and assembly QC</li>
        <li class="dropdown-header">Genome annotation</li>
        <li class="dropdown-header">Pangenomics &amp; phylogenetics</li>
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./homework.html"> 
<span class="menu-text">Homework</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-reference-pages" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Reference Pages</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-reference-pages">    
        <li>
    <a class="dropdown-item" href="./open_shell.html">
 <span class="dropdown-text">Opening a Unix shell at OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./rstudio.html">
 <span class="dropdown-text">Starting an RStudio session at OSC</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="./ref_shell2.html">
 <span class="dropdown-text">Continued Unix shell intro</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./ref_shell.html">
 <span class="dropdown-text">Unix shell overview/reference</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="./ref_software.html">
 <span class="dropdown-text">Software management</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./ref_data.html">
 <span class="dropdown-text">Data management</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/cabana-workshop">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/cabana-workshop/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#interpreting-fastqc-output" id="toc-interpreting-fastqc-output" class="nav-link" data-scroll-target="#interpreting-fastqc-output"><span class="header-section-number">1</span> Interpreting FastQC output</a>
  <ul class="collapse">
  <li><a href="#fastqc-html-modules" id="toc-fastqc-html-modules" class="nav-link" data-scroll-target="#fastqc-html-modules"><span class="header-section-number">1.1</span> FastQC HTML modules</a></li>
  <li><a href="#interpreting-our-fastqc-results" id="toc-interpreting-our-fastqc-results" class="nav-link" data-scroll-target="#interpreting-our-fastqc-results"><span class="header-section-number">1.2</span> Interpreting our FastQC results</a></li>
  </ul></li>
  <li><a href="#multiqc" id="toc-multiqc" class="nav-link" data-scroll-target="#multiqc"><span class="header-section-number">2</span> MultiQC</a>
  <ul class="collapse">
  <li><a href="#running-multiqc" id="toc-running-multiqc" class="nav-link" data-scroll-target="#running-multiqc"><span class="header-section-number">2.1</span> Running MultiQC</a></li>
  <li><a href="#multiqc-output" id="toc-multiqc-output" class="nav-link" data-scroll-target="#multiqc-output"><span class="header-section-number">2.2</span> MultiQC output</a></li>
  </ul></li>
  <li><a href="#trimgalore" id="toc-trimgalore" class="nav-link" data-scroll-target="#trimgalore"><span class="header-section-number">3</span> TrimGalore</a>
  <ul class="collapse">
  <li><a href="#other-trimgalore-options" id="toc-other-trimgalore-options" class="nav-link" data-scroll-target="#other-trimgalore-options"><span class="header-section-number">3.1</span> Other TrimGalore options</a></li>
  <li><a href="#trimgalore-output" id="toc-trimgalore-output" class="nav-link" data-scroll-target="#trimgalore-output"><span class="header-section-number">3.2</span> TrimGalore output</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Read preprocessing</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 8, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p><br></p>
<section id="introduction" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="introduction">Introduction</h2>
<p>The first series of steps our bacterial whole-genome analysis workflow concerns the “preprocessing” of the reads.</p>
<p>In part, that is just quality control (QC), which will leave the data untouched, but it also entails the removal of unwanted bits of sequence. After preprocessing, we will still have FASTQ files, just with somewhat less content.</p>
<p>We will preprocess our reads with the following steps:</p>
<ol type="1">
<li>QC with <strong>FastQC</strong> – we already ran that yesterday, but will interpret the results now</li>
<li>Summarizing FastQC results with <strong>MultiQC</strong></li>
<li>Removing adapters and low-quality bases from our reads with <strong>TrimGalore</strong></li>
</ol>
<section id="setting-up" class="level4 exercise">
<h4 class="anchored" data-anchor-id="setting-up"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Setting up</strong></h4>
<p>You should have an active VS Code session with an open terminal. In that terminal, you should be be in your dir <code>/fs/scratch/PAS2250/cabana/$USER/bact</code>.</p>
</section>
<p><br></p>
</section>
<section id="interpreting-fastqc-output" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="interpreting-fastqc-output"><span class="header-section-number">1</span> Interpreting FastQC output</h2>
<section id="fastqc-html-modules" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="fastqc-html-modules"><span class="header-section-number">1.1</span> FastQC HTML modules</h3>
<p>We’ll now go through a couple of the FastQC plots/modules, with first some example plots<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> with good/bad results for reference.</p>
<section id="overview-of-module-results" class="level4">
<h4 class="anchored" data-anchor-id="overview-of-module-results">Overview of module results</h4>
<p>FastQC has “<strong>pass</strong>” (checkmark in green), “<strong>warning</strong>” (exclamation mark in orange), and “<strong>fail</strong>” (cross in red) assessments for each module, as you can see below.</p>
<p>These are handy and typically at least somewhat meaningful, but it is important to realize <em>that a “warning” or a “fail” is not necessarily the bad news that it may appear to be,</em> because, e.g.:</p>
<ul>
<li>Some of these modules could perhaps be called overly strict.</li>
<li>Some warnings and fails are easily remedied or simply not a very big deal.</li>
<li>FastQC assumes that your data is derived from whole-genome shotgun sequencing — some other types of data like RNA-seq data will always trigger a couple of warnings and files based on expected differences.</li>
</ul>
<p align="center">
</p><p><img src="img/fastqc_summary2.png" width="30%"></p>
<p></p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="basic-statistics" class="level4">
<h4 class="anchored" data-anchor-id="basic-statistics">Basic statistics</h4>
<p>This shows, for example, the number of sequences (reads) and the read length range for your file:</p>
<p align="center">
</p><p><img src="img/fastqc_mod01_basic-stats.png" width="85%"></p>
<p></p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="per-base-quality-sequence-quality" class="level4">
<h4 class="anchored" data-anchor-id="per-base-quality-sequence-quality">Per base quality sequence quality</h4>
<p>This figure visualize the mean per-base quality score (y-axis) along the length of the reads (x-axis). Note that:</p>
<ul>
<li>A decrease in sequence quality along the reads is normal.</li>
<li>R2 (reverse) reads are usually worse than R1 (forward) reads.</li>
</ul>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Good / acceptable:</strong></p>
<p align="center">
</p><p><img src="img/fastqc_good.png" width="100%"></p>
<p></p>
</div><div class="column" style="width:50%;">
<p><strong>Bad:</strong></p>
<p align="center">
</p><p><img src="img/fastqc_bad.png" width="100%"></p>
<p></p>
</div>
</div>
<p>To interpret the quality scores along the y-axis, note the color scaling in the graphs (green is good, etc.), and see this table for details:</p>
<table class="table">
<thead>
<tr class="header">
<th>Phred quality score</th>
<th>Error probability</th>
<th>Rough interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>10</strong></td>
<td>1 in 10</td>
<td>terrible</td>
</tr>
<tr class="even">
<td><strong>20</strong></td>
<td>1 in 100</td>
<td>bad</td>
</tr>
<tr class="odd">
<td><strong>30</strong></td>
<td>1 in 1,000</td>
<td>good</td>
</tr>
<tr class="even">
<td><strong>40</strong></td>
<td>1 in 10,000</td>
<td>excellent</td>
</tr>
</tbody>
</table>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="per-sequence-quality-scores" class="level4">
<h4 class="anchored" data-anchor-id="per-sequence-quality-scores">Per sequence quality scores</h4>
<p>This shows the same quality scores we saw above, but now simply as a density plot of per-read averages, with the quality score now along the x-axis, and the number of reads with that quality score along the y-axis:</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Good:</strong></p>
<p align="center">
</p><p><img src="img/fastqc_mod04_per-seq-qual_good.png" width="100%"></p>
<p></p>
</div><div class="column" style="width:50%;">
<p><strong>Bad:</strong></p>
<p align="center">
</p><p><img src="img/fastqc_mod04_per-seq-qual_bad.png" width="100%"></p>
<p></p>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="sequence-length-distribution" class="level4">
<h4 class="anchored" data-anchor-id="sequence-length-distribution">Sequence length distribution</h4>
<p>Will throw a warning as soon as not all sequences are of the same length (like below), but this is quite normal.</p>
<p align="center">
</p><p><img src="img/fastqc_mod08_seqlen_warning.png" width="60%"></p>
<p></p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="adapter-content" class="level4">
<h4 class="anchored" data-anchor-id="adapter-content">Adapter content</h4>
<p>Checks for known adapter sequences. When some of the insert sizes are shorter than the read length, adapters can end up in the sequence – these should be removed!</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Good:</strong></p>
<p align="center">
</p><p><img src="img/fastqc_mod11_adapter_good.png" width="100%"></p>
<p></p>
</div><div class="column" style="width:50%;">
<p><strong>Bad:</strong></p>
<p align="center">
</p><p><img src="img/fastqc_mod11_adapter_bad.png" width="100%"></p>
<p></p>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
</section>
<section id="interpreting-our-fastqc-results" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="interpreting-our-fastqc-results"><span class="header-section-number">1.2</span> Interpreting our FastQC results</h3>
<p>First, we’ll unfortunately have to download FastQC’s output HTML files<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>:</p>
<ul>
<li>Find the FastQC HTML files in the file explorer in the VS Code side bar.</li>
<li>Right-click on one of them, click <code>Download...</code> and follow the prompt to download the file somewhere to your computer (doesn’t matter where, just make sure you see where it goes).</li>
<li>Repeat this for the second file</li>
<li>Then, open your computer’s file browser, find the downloaded files, and double-click on one. It should be <strong>opened in your default web browser</strong>.</li>
</ul>
<section id="exercise-interpreting-our-fastqc-results" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-interpreting-our-fastqc-results"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Interpreting our FastQC results</h4>
<ul>
<li><p>Open the HTML file for the R1 FASTQ file and go through the modules we discussed above. Can you make sense of it? Does the data look good to you, overall?</p></li>
<li><p>Now open the HTML file for the <strong>R2</strong> FASTQ file and take a look just at the quality scores. Does it look any worse than the R1?</p></li>
</ul>
</section>
<p><br></p>
</section>
</section>
<section id="multiqc" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="multiqc"><span class="header-section-number">2</span> MultiQC</h2>
<p>Here are some challenges you may run into after running FastQC:</p>
<ul>
<li><p>When you have many FASTQ files, you’ll generate a lot of FastQC HTML files to sort through. Our dataset is small with only 16 samples, but this still means 32 FastQC outputs. Other datasets may easily have dozens or even hundreds of samples, in which case checking all of the output becomes a very unpleasant task.</p></li>
<li><p>Even if you do diligently go through each file, it’s not that easy to compare samples, since they are not drawn in the same graphs.</p></li>
</ul>
<p>MultiQC addresses these problems as it aggregates FastQC results from many files, and summarizes them into a single HTML file with (still) one graph per FastQC module.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Not just for FastQC results!
</div>
</div>
<div class="callout-body-container callout-body">
<p>And while MultiQC is most widely used for FastQC aggregation, it can recognize and process the (often “log”-type) output of dozens of bioinformatics tools, including several others that we will be using.</p>
</div>
</div>
<p>MultiQC’s graphs are also <em>interactive</em>, but here is a static example:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_fastqc_per_base_sequence_quality_plot.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:95.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<section id="running-multiqc" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="running-multiqc"><span class="header-section-number">2.1</span> Running MultiQC</h3>
<p>To run MultiQC, use the command <code>multiqc</code>. Let’s start by running it with the <code>--help</code> option:</p>
<details>
<summary>
You should still have the <code>cabana</code> Conda environment active. If not, click here for instructions
</summary>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the Conda enviroment</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/scratch/PAS2250/jelmer/cabana</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--help</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (Only the top part of the output is shown in the screenshot below)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_help.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
<p>As the first couple of help lines in the paler gray color explain, MultiQC will search the <code>[ANALYSIS DIRECTORY]</code>, a dir that we pass to it as an argument at the end of the command line.</p>
<p>That is, if we tell MultiQC about the <code>results/fastqc_pretrim</code> directory like so, it should find and then aggregate all the FastQC results in there:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this - we'll complete the command in a second)</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> ../../bact_results/fastqc_pretrim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Since you ran FastQC on only sample, we’ll be using the FastQC results for all files that I generated for you.
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>The default output directory of MultiQC is the current working directory, so just like with FastQC, we do want to use that option as well<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run MultiQC to summarize the FastQC results</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outdir</span> results/multiqc_fastqc <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  ../../bact_results/fastqc_pretrim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_stdout.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="multiqc-output" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="multiqc-output"><span class="header-section-number">2.2</span> MultiQC output</h3>
<p>Then, you should have some files in the output dir:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/multiqc_fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 1.7M
drwxr-xr-x 2 jelmer PAS2250 4.0K Feb  4 14:57 multiqc_data
-rw-r--r-- 1 jelmer PAS2250 1.7M Feb  4 14:57 multiqc_report.html</code></pre>
<p>Go ahead and find the HTML file in VS Code’s file browser, right-click on it and then download it to your computer, and click on the file in your own computer to open it in your browser (i.e., just like we did with the FastQC output).</p>
<section id="exercise-explore-the-multiqc-results" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-explore-the-multiqc-results"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Explore the MultiQC results</h4>
<p>Check for example whether patterns are consistent across samples, or if there are any outliers.</p>
</section>
<p><br></p>
</section>
</section>
<section id="trimgalore" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="trimgalore"><span class="header-section-number">3</span> TrimGalore</h2>
<p>We will run TrimGalore to filter our FASTQ files, removing:</p>
<ul>
<li>Any adapter sequences that may be present in the reads</li>
<li>Poor-quality bases at the start and end of the reads</li>
<li>Very short reads (in most cases made short by the prior two steps)</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why TrimGalore, specifically? <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Several largely equivalent tools exist for this kind of FASTQ preprocessing — <em>Trimmomatic</em> and <em>fastp</em> are two other commonly used ones. (And TrimGalore itself is mostly a wrapper around another tool called <em>CutAdapt</em>.)</p>
<p>Two advantages of of TrimGalore are that it will <strong>auto-detect the adapters</strong> that are present in your reads (e.g., different library prep protocols use different adapters), and that it can <strong>automatically run <em>FastQC</em></strong> on the processed sequences.</p>
</div>
</div>
</div>
<p>Unfortunately, we’ll first have to switch Conda environments for TrimGalore:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/ess/PAS0471/jelmer/conda/trimgalore</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
<hr style="height:1pt; visibility:hidden;">
<section id="paired-end-trimming" class="level4">
<h4 class="anchored" data-anchor-id="paired-end-trimming">Paired-end trimming</h4>
<p>From the line below “USAGE:”, we can tell that the FASTQ file name(s) should be specified as <em>positional argument(s) at the end of the command</em>.</p>
<p><strong>We will run TrimGalore for one sample at a time</strong>, but because we have paired-end reads:</p>
<ul>
<li><p>We’ll specify two FASTQ file names: one with the forward (R1) reads, and one with the reverse (R2) reads.</p></li>
<li><p>We’ll have to use the <code>--paired</code> option (otherwise, TrimGalore will only process the R1 and R2 files separately, and omit the final step where it removes orphaned reads — see the box below for details).</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More on paired-end trimming <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>When we have paired-end reads, much of the trimming happens separately for the R1 (forward) and R2 (reverse) files, but at the end of the run, TrimGalore will make sure that every R1 read still has its R2 counterpart, and vice versa.</p>
<p>Any “orphaned” reads will by default be removed, because R1 and R2 files for the same samples always need to contain all the same reads. (TrimGalore does have an option to retain these orphaned reads into <em>separate</em> files, but we won’t use that.)</p>
</div>
</div>
</div>
<p>So far, our command looks like this:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this)</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--paired</span> <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  data/fastq/SM04_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  data/fastq/SM04_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="other-trimgalore-options" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="other-trimgalore-options"><span class="header-section-number">3.1</span> Other TrimGalore options</h3>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Checking the TrimGalore help
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You might also want to run TrimGalore with the <code>--help</code> option to learn how to run it:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="at">--help</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: Below I am only showing (truncated) output for the key options!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code> USAGE:
trim_galore [options] &lt;filename(s)&gt;

--paired                This option performs length trimming of quality/adapter/RRBS trimmed reads for
                        paired-end files.
-o/--output_dir &lt;DIR&gt;   If specified all output will be written to this directory instead of the current
                        directory. If the directory doesn't exist it will be created for you.
--fastqc                Run FastQC in the default mode on the FastQ file once trimming is complete.
--fastqc_args "&lt;ARGS&gt;"  Passes extra arguments to FastQC.
-a/--adapter &lt;STRING&gt;   Adapter sequence to be trimmed. If not specified explicitly, Trim Galore will
                        try to auto-detect whether the Illumina universal, Nextera transposase or Illumina
                        small RNA adapter sequence was used.
-q/--quality &lt;INT&gt;      Trim low-quality ends from reads in addition to adapter removal.
--length &lt;INT&gt;          Discard reads that became shorter than length INT because of either
                        quality or adapter trimming. A value of '0' effectively disables
                        this behaviour. Default: 20 bp.</code></pre>
</div>
</div>
</div>
<p>We will:</p>
<ul>
<li>Use the default settings for adapters (auto-detection and removal) and the base quality threshold (a Phred score of 20)</li>
<li>Use a longer read length (36 bp) than the default of 36 bp: <strong><code>--length 36</code></strong>.</li>
<li>Specify the output directory<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>: <strong><code>--output_dir results/trimgalore</code></strong>.</li>
<li>Have TrimGalore run <em>FastQC</em> for us on the filtered FASTQ files, so we can e.g.&nbsp;check if adapters were successfully removed.</li>
</ul>
<p>A final test command to run TrimGalore on our actual (but small, subsetted) FASTQ files in <code>data/fastq</code> could therefore look as follows:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a> <span class="co"># Once again, we have to make the FastQC outdir!</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results/fastqc_posttrim</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run TrimGalore</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="ex">trim_galore</span> <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--paired</span> <span class="dt">\</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--length</span> 36 <span class="dt">\</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--output_dir</span> results/trimgalore <span class="dt">\</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastqc_args</span> <span class="st">"--outdir results/fastqc_posttrim"</span> <span class="dt">\</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    data/fastq/SM04_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    data/fastq/SM04_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Multicore support not enabled. Proceeding with single-core trimming.
Path to Cutadapt set as: 'cutadapt' (default)
Cutadapt seems to be working fine (tested command 'cutadapt --version')
Cutadapt version: 4.4
single-core operation.
igzip command line interface 2.30.0
igzip detected. Using igzip for decompressing

No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)

Output directory results/trimgalore/ doesn't exist, creating it for you...

Output will be written into the directory: /fs/scratch/PAS2250/cabana/jelmer/bact/results/trimgalore/

AUTO-DETECTING ADAPTER TYPE
===========================
Attempting to auto-detect adapter type from the first 1 million sequences of the first file (&gt;&gt; data/fastq/SM04_R1.fastq.gz &lt;&lt;)
# [...output truncated...]</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Trimming strictness
</div>
</div>
<div class="callout-body-container callout-body">
<p>The exact choice of trimming strictness parameters, especially regarding the minimum read length and minimum base quality is fairly arbitrary, and may not have a large effect on the final assembly. That said, while read lengths of 20 may still be useful in a read mapping context, they are less so in a de novo assembly context, so we use a higher read length threshold.</p>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="trimgalore-output" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="trimgalore-output"><span class="header-section-number">3.2</span> TrimGalore output</h3>
<p>After you ran the command above, a lot of logging output should have been printed to screen.</p>
<p>For example, it reports the adapter that it detected, the final parameters passed to Cutadapt (which does the actual trimming), and results on how much sequence was removed.</p>
<section id="exercise-check-the-trimgalore-logging-output" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-check-the-trimgalore-logging-output"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Check the TrimGalore logging output</h4>
<p>Look for <code>=== Summary ===</code> sections (two of them, one for the R1 and one for the R2) to answer the following questions:</p>
<ul>
<li>What percentage of the reads had adapter sequences?</li>
<li>What percentage of basepairs were quality-trimmed from the R1 and R2 files, respectively?</li>
</ul>
<p>Look near the end of the output to answer the following question:</p>
<ul>
<li>How many reads were removed due to the length-filter?</li>
</ul>
<details>
<summary>
<em>Click for the solution</em>
</summary>
<p>Some of the more relevant information that should have been printed:</p>
<ul>
<li>For the R1 file:</li>
</ul>
<pre class="bash-out"><code>=== Summary ===

Total reads processed:               1,031,129
Reads with adapters:                   398,726 (38.7%)
Reads written (passing filters):     1,031,129 (100.0%)

Total basepairs processed:   275,173,541 bp
Quality-trimmed:               7,733,000 bp (2.8%)
Total written (filtered):    266,857,947 bp (97.0%)</code></pre>
<ul>
<li>For the R2 file:</li>
</ul>
<pre class="bash-out"><code>=== Summary ===

Total reads processed:               1,031,129
Reads with adapters:                   424,883 (41.2%)
Reads written (passing filters):     1,031,129 (100.0%)

Total basepairs processed:   276,647,678 bp
Quality-trimmed:              42,814,676 bp (15.5%)
Total written (filtered):    233,141,370 bp (84.3%)</code></pre>
<ul>
<li>And at the end:</li>
</ul>
<pre class="bash-out"><code>Number of sequence pairs removed because at least one read was shorter than the length cutoff (36 bp): 6672 (0.65%)</code></pre>
</details>
</section>
<hr style="height:1pt; visibility:hidden;">
<p>Much of the above information is also saved in the output dir in files that end in <code>*_trimming_report.txt</code>.</p>
<p>The main output files, however, are a new pair of FASTQ files with trimmed reads — we will use those FASTQ files for the next step (assembly). Let’s take a look:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/trimgalore</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 342M
-rw-r--r-- 1 jelmer PAS0471 5.2K Feb  4 16:48 SM04_R1.fastq.gz_trimming_report.txt
-rw-r--r-- 1 jelmer PAS0471 169M Feb  4 16:52 SM04_R1_val_1.fq.gz
-rw-r--r-- 1 jelmer PAS0471 5.2K Feb  4 16:52 SM04_R2.fastq.gz_trimming_report.txt
-rw-r--r-- 1 jelmer PAS0471 174M Feb  4 16:52 SM04_R2_val_2.fq.gz</code></pre>
<p>Also, we should have FastQC output files that will be good to check:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/fastqc_posttrim</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 2.4M
-rw-r--r-- 1 jelmer PAS0471 705K Feb  4 17:00 SM04_R1_val_1_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 437K Feb  4 17:00 SM04_R1_val_1_fastqc.zip
-rw-r--r-- 1 jelmer PAS0471 724K Feb  4 17:01 SM04_R2_val_2_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 470K Feb  4 17:01 SM04_R2_val_2_fastqc.zip</code></pre>
<section id="exercise-check-the-trimming-results-with-fastqc" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-check-the-trimming-results-with-fastqc"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Check the trimming results with FastQC</h4>
<ul>
<li>Check the FastQC output HTML to see if adapters were removed, and if the overall quality looks better.</li>
</ul>
</section>
<p><br></p>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p> Attribution: Some of the FastQC example plots were taken from <a href="https://rtsf.natsci.msu.edu/sites/_rtsf/assets/File/FastQC_TutorialAndFAQ_080717.pdf">here</a>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>The installed version of VS Code does not allow us to view HTML files<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p> I will specify the dir <code>results/multiqc_fastqc</code>, indicating that this is a MultiQC run that summarizes FastQC output<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>Once again, the default output dir is the current working dir which is not convenient<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>